### DISCUSSION TITLE
Is there any way to switch to a pending ostree deployment WITHOUT REBOOTING

### DISCUSSION THREAD

**Discussion by gwaines** [ostreedev/ostree#3336] [Q&A]
Is there any way to switch to a pending ostree deployment WITHOUT REBOOTING ?
... for the scenario where you know that you do not need a reboot.

For example,
you can deploy the commit

ostree admin deploy f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1
	Copying /etc changes: 91 modified, 27 removed, 335 added
	Transaction complete; bootconfig swap: no; bootversion: boot.0.0, deployment count change: 0

... which creates the deployment in a 'pending' state.

ostree admin status
	  debian f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1.0 (pending)
	    origin refspec: f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1
	* debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1
    origin refspec: debian:starlingx

And you can swap around the bind mounts for /etc and /usr ... to switch /usr and /etc to point to this pending deployment
e.g.
	OROOT=`ostree admin status |grep pending  |awk '{print $2}'`
	echo $OROOT
	    f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1.0
	NROOT=/ostree/deploy/debian/deploy/$OROOT
	echo $NROOT
	    /ostree/deploy/debian/deploy/f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1.0

        umount -l /usr /etc  
        mount --bind -o ro,noatime $NROOT/usr /usr
        mount --bind -o ro,noatime $NROOT/etc /etc 

But the 'ostree admin status' is still showing the deployment as 'pending' ... 

ostree admin status
	  debian f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1.0 (pending)
	    origin refspec: f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1
	* debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1
    origin refspec: debian:starlingx

... which causes problems on subsequent 'ostree admin deploy ...' commands;
i.e. specifically the 3-way /etc merge because ostree doesn't think the 'pending' deployment is 'active'.

Any ideas ?

**Comment by prydom** [comment]
If you're booted from the deployment you can use `systemctl stop ostree-finalize-staged.service` to trigger the existing staged deployment. That's the same trigger that is used to trigger the 3-way merge at power off.

Internally what this does is execute `ostree admin finalize-staged`. Note that any changes made to /etc after the 3-way merge would be only in your fallback deployment.

**Reply by gwaines** [reply]
Hey @prydom 

I tried this ... but it did not seem to change the state of the 'pending' deployment to active('*'), which is what causes the problems on the subsequent 'ostree admin deploy ...' commands.

e.g.

// pull a new commit into my sysroot ostree
% ostree --repo=/sysroot/ostree/repo pull debian starlingx --depth=-1
6 metadata, 3 content objects imported; 336 B transferred in 8 seconds; 802.6 kB content written

// deploy it
% ostree admin deploy starlingx --no-prune --retain
Copying /etc changes: 91 modified, 27 removed, 335 added
Transaction complete; bootconfig swap: yes; bootversion: boot.1.1, deployment count change: 1

// ... so new deployment is in 'pending' state
% ostree admin status 
  debian 4b0d4ee47f52238b0e036eda3906c0daa4de6a7eb312474ee367a1dbb8cdaebe.0 (pending)
    origin refspec: starlingx
\* debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1
    origin refspec: debian:starlingx
  debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.0 (rollback)
    origin refspec: debian:starlingx

// I still swapped around the bind mounts for /etc and /usr ... to switch /usr and /etc to point to this pending deployment
OROOT=ostree admin status |grep pending  |awk '{print $2}'
echo $OROOT
f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1.0
NROOT=/ostree/deploy/debian/deploy/$OROOT
echo $NROOT
/ostree/deploy/debian/deploy/f5982e1314fb042faf75759ec5cbfb8f0838e31a78b03f13bdb0c54653edbed1.0

// then when i try to do 'systemctl stop ostree-finalize-staged.service' 
% systemctl stop ostree-finalize-staged.service
Warning: Stopping ostree-finalize-staged.service, but it can still be activated by:
  ostree-finalize-staged.path

// but the 'pending' deployment is still 'pending'  :(
% ostree admin status
  debian 4b0d4ee47f52238b0e036eda3906c0daa4de6a7eb312474ee367a1dbb8cdaebe.0 (pending)
    origin refspec: starlingx
\* debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1
    origin refspec: debian:starlingx
  debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.0 (rollback)
    origin refspec: debian:starlingx



??? Any Ideas ???

Do I need to create a 'staged' deployment to do this ?
e.g.
ostree admin upgrade --stage ...
instead of 
ostree admin deploy ...

OR 

run 'ostree admin deploy ...'  with '--stage' option ?

???

**Reply by gwaines** [reply]
Actually I tried using '--stage' option with 'ostree admin deploy ...'
but that just seemed to have the deployment go thru a 'staged' state before getting to a 'pending' state again

i.e.
% ostree admin deploy starlingx --no-prune --retain --stage
%
% ostree admin status
  debian 2f4107ffb306719741a2a2a929b50ed3d82907ad2ffa6101224994faf1b020a0.0 (staged)
    origin refspec: starlingx
\* debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1
    origin refspec: debian:starlingx
  debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.0 (rollback)
    origin refspec: debian:starlingx
%
% systemctl stop ostree-finalize-staged.service
Warning: Stopping ostree-finalize-staged.service, but it can still be activated by:
  ostree-finalize-staged.path
%
%
% ostree admin status
  debian 2f4107ffb306719741a2a2a929b50ed3d82907ad2ffa6101224994faf1b020a0.0 (pending)
    origin refspec: starlingx
\* debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1
    origin refspec: debian:starlingx
%


I want to get the deployment into the active('*') state ... without having to go thru a reboot.
e.g.
% ostree admin status
\* debian 2f4107ffb306719741a2a2a929b50ed3d82907ad2ffa6101224994faf1b020a0.0
    origin refspec: starlingx
  debian ce335da86a8b03d36f6041f46b7d86754d7f2c97e29114d91742abb6f72ee8fd.1 (rollback)
    origin refspec: debian:starlingx


I need ostree to believe that 2f4107ffb306719741a2a2a929b50ed3d82907ad2ffa6101224994faf1b020a0.0 deployment is the active ('*') deployment so the NEXT 'ostree admin deploy' command's etc merge will work correctly.

**Reply by prydom** [reply]
Ah so you want something like `rpm-ostree apply-live --allow-replacement` but for Debian? I'm not sure such a tool exists currently.

Take a look at https://github.com/coreos/rpm-ostree/blob/2521d5651725b65694a9caac091369989922e629/rust/src/live.rs#L333 to see what that would take. I believe the state for the current boot is serialized into /run/ostree-booted and some files in /run/ostree.
